From 0365b1f803ef181d5f806b3a8b77afe6c5edbf91 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@gnome.org>
Date: Wed, 7 Oct 2015 01:21:02 +0200
Subject: [PATCH] st-button: Don't rely on hover to accept clicks

Since commit 4f1f226828 we only consider buttons clicked when the
release event had a corresponding press event. However as we use
the hover state to check whether a release event actually occurred
on the button, we dismiss any clicks in cases where we missed the
enter event - most likely due to some other actor holding a grab.
Instead, check whether the button contains the event's source, which
should be less error-prone.

https://bugzilla.gnome.org/show_bug.cgi?id=748919
---
 src/st/st-button.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/st/st-button.c b/src/st/st-button.c
index 56502f1..0db6b0d 100644
--- a/src/st/st-button.c
+++ b/src/st/st-button.c
@@ -211,7 +211,7 @@ st_button_button_release (ClutterActor       *actor,
     {
       gboolean is_click;
 
-      is_click = button->priv->grabbed && st_widget_get_hover (ST_WIDGET (button));
+      is_click = button->priv->grabbed && clutter_actor_contains (actor, event->source);
       st_button_release (button, device, mask, is_click ? event->button : 0, NULL);
 
       button->priv->grabbed &= ~mask;
-- 
2.7.2

